File: email_utils/email_parser.py
import base64
import logging
import email
from email.header import decode_header

logger = logging.getLogger(__name__)

def decode_str(s):
    """Decodes a possibly encoded string header."""
    if s is None:
        return ""
    try:
        decoded_parts = decode_header(s)
        decoded_str = ""
        for part, encoding in decoded_parts:
            if isinstance(part, bytes):
                if encoding:
                    try:
                        decoded_str += part.decode(encoding)
                    except Exception:
                        decoded_str += part.decode("utf-8", errors="replace")
                else:
                    decoded_str += part.decode("utf-8", errors="replace")
            else:
                decoded_str += part
        return decoded_str
    except Exception as e:
        logger.error(f"Error decoding string: {e}")
        return s

def get_body_from_message(message):
    """Extracts the body text from a MIME message."""
    if message.is_multipart():
        # Try plain text parts first.
        for part in message.walk():
            content_type = part.get_content_type()
            content_disposition = str(part.get("Content-Disposition"))
            if "attachment" in content_disposition:
                continue
            if content_type == "text/plain":
                try:
                    return part.get_payload(decode=True).decode("utf-8", errors="replace")
                except Exception as e:
                    logger.error(f"Error decoding plain text part: {e}")
        # Then try HTML parts.
        for part in message.walk():
            content_type = part.get_content_type()
            content_disposition = str(part.get("Content-Disposition"))
            if "attachment" in content_disposition:
                continue
            if content_type == "text/html":
                try:
                    return part.get_payload(decode=True).decode("utf-8", errors="replace")
                except Exception as e:
                    logger.error(f"Error decoding HTML part: {e}")
    else:
        try:
            body = message.get_payload(decode=True)
            if body:
                return body.decode("utf-8", errors="replace")
        except Exception as e:
            logger.error(f"Error decoding non-multipart message: {e}")
    return "No readable content found"

def parse_email_content(email_data):
    """
    Parses email data from the Gmail API and extracts sender, subject, and body.
    """
    try:
        payload = email_data.get("payload", {})
        headers = payload.get("headers", [])
        sender = next((h["value"] for h in headers if h["name"].lower() == "from"), "Unknown Sender")
        subject = next((h["value"] for h in headers if h["name"].lower() == "subject"), "No Subject")
        sender = decode_str(sender)
        subject = decode_str(subject)

        body = ""
        if "body" in payload and "data" in payload["body"]:
            try:
                body = base64.urlsafe_b64decode(payload["body"]["data"]).decode("utf-8", errors="replace")
            except Exception as e:
                logger.error(f"Error decoding body data: {e}")
        if not body and "parts" in payload:
            for part in payload["parts"]:
                if part.get("mimeType") == "text/plain" and "body" in part and "data" in part["body"]:
                    try:
                        body = base64.urlsafe_b64decode(part["body"]["data"]).decode("utf-8", errors="replace")
                        break
                    except Exception as e:
                        logger.error(f"Error decoding part data: {e}")
        if not body and "parts" in payload:
            for part in payload["parts"]:
                if part.get("mimeType") == "text/html" and "body" in part and "data" in part["body"]:
                    try:
                        body = base64.urlsafe_b64decode(part["body"]["data"]).decode("utf-8", errors="replace")
                        break
                    except Exception as e:
                        logger.error(f"Error decoding HTML part data: {e}")
        body = body.strip()
        return {"sender": sender, "subject": subject, "body": body}
    except Exception as e:
        logger.exception(f"Error parsing email: {e}")
        return None

File: backend/emailParser.ts
import { Buffer } from "node:buffer";
import type { gmail_v1 } from "googleapis";
import { PDFParse } from "pdf-parse";
import type { GmailMessage, GmailMessagePart, ParsedEmail, ParsedEmailAttachment } from "./types";
import { logger } from "./logger";

function decodeMimeWords(value: string) {
  return value.replace(/=\?([^?]+)\?([BQbq])\?([^?]+)\?=/g, (_match, charset, encoding, text) => {
    try {
      const bytes =
        String(encoding).toUpperCase() === "B"
          ? Buffer.from(text, "base64")
          : Buffer.from(String(text).replace(/_/g, " ").replace(/=([0-9A-Fa-f]{2})/g, (_m, hex) => String.fromCharCode(Number.parseInt(hex, 16))), "binary");
      return bytes.toString(String(charset).toLowerCase() as BufferEncoding);
    } catch {
      return text;
    }
  });
}

function base64UrlDecode(data: string) {
  const normalized = data.replace(/-/g, "+").replace(/_/g, "/");
  return Buffer.from(normalized, "base64").toString("utf8").trim();
}

function base64UrlDecodeBuffer(data: string) {
  const normalized = data.replace(/-/g, "+").replace(/_/g, "/");
  return Buffer.from(normalized, "base64");
}

function headerValue(message: GmailMessage, name: string, fallback: string) {
  const headers = message.payload?.headers ?? [];
  const header = headers.find((candidate) => candidate.name?.toLowerCase() === name);
  return decodeMimeWords(header?.value ?? fallback);
}

function isAttachment(part: GmailMessagePart) {
  const disposition = part.headers?.find(
    (header) => header.name?.toLowerCase() === "content-disposition",
  )?.value;
  return disposition?.toLowerCase().includes("attachment") ?? Boolean(part.filename);
}

function collectAttachmentParts(part: GmailMessagePart | null | undefined, result: GmailMessagePart[] = []) {
  if (!part) return result;
  if (isAttachment(part) && part.body?.attachmentId) {
    result.push(part);
  }
  for (const child of part.parts ?? []) {
    collectAttachmentParts(child, result);
  }
  return result;
}

function findBody(part: GmailMessagePart | null | undefined, mimeType: string): string | null {
  if (!part || isAttachment(part)) {
    return null;
  }

  if (part.mimeType === mimeType && part.body?.data) {
    try {
      return base64UrlDecode(part.body.data);
    } catch (error) {
      logger.error(`Error decoding ${mimeType} body`, error instanceof Error ? error.message : error);
    }
  }

  for (const child of part.parts ?? []) {
    const body = findBody(child, mimeType);
    if (body) {
      return body;
    }
  }
  return null;
}

export function parseSender(sender: string) {
  const match = sender.match(/<([^>]+)>/);
  return {
    fromName: sender.replace(/<[^>]+>/g, "").replace(/^"|"$/g, "").trim() || sender || "Unknown",
    fromEmail: match?.[1] ?? "",
  };
}

async function extractPdfText(data: Buffer, filename: string) {
  const parser = new PDFParse({ data });
  try {
    const result = await parser.getText({ first: 8 });
    return result.text.trim();
  } catch (error) {
    logger.error(`Error extracting PDF text from ${filename}`, error instanceof Error ? error.message : error);
    return "";
  } finally {
    await parser.destroy();
  }
}

async function extractAttachmentText(
  service: gmail_v1.Gmail,
  messageId: string,
  part: GmailMessagePart,
): Promise<ParsedEmailAttachment | null> {
  const attachmentId = part.body?.attachmentId;
  const filename = decodeMimeWords(part.filename || "attachment");
  const mimeType = part.mimeType || "application/octet-stream";
  if (!attachmentId) return null;

  try {
    const response = await service.users.messages.attachments.get({
      userId: "me",
      messageId,
      id: attachmentId,
    });
    const data = response.data.data ? base64UrlDecodeBuffer(response.data.data) : Buffer.alloc(0);
    if (data.length === 0) return null;

    let text = "";
    if (mimeType === "application/pdf" || filename.toLowerCase().endsWith(".pdf")) {
      text = await extractPdfText(data, filename);
    } else if (
      mimeType.startsWith("text/") ||
      /\.(txt|md|csv|json|log)$/i.test(filename)
    ) {
      text = data.toString("utf8").trim();
    }

    if (!text) {
      return null;
    }

    return {
      filename,
      mimeType,
      text: text.length > 24_000 ? `${text.slice(0, 24_000)}\n[attachment text truncated]` : text,
    };
  } catch (error) {
    logger.error(`Error loading attachment ${filename}`, error instanceof Error ? error.message : error);
    return null;
  }
}

export async function extractEmailAttachments(
  service: gmail_v1.Gmail,
  emailData: GmailMessage,
): Promise<ParsedEmailAttachment[]> {
  const messageId = emailData.id ?? "";
  if (!messageId) return [];
  const parts = collectAttachmentParts(emailData.payload).slice(0, 4);
  const attachments: ParsedEmailAttachment[] = [];

  for (const part of parts) {
    const attachment = await extractAttachmentText(service, messageId, part);
    if (attachment) {
      attachments.push(attachment);
    }
  }

  return attachments;
}

export function parseEmailContent(emailData: GmailMessage): ParsedEmail | null {
  try {
    const sender = headerValue(emailData, "from", "Unknown Sender");
    const subject = headerValue(emailData, "subject", "No Subject");
    const body =
      findBody(emailData.payload, "text/plain") ??
      findBody(emailData.payload, "text/html") ??
      (emailData.payload?.body?.data ? base64UrlDecode(emailData.payload.body.data) : "") ??
      "";

    return {
      sender,
      subject,
      body: body.trim(),
      messageId: emailData.id ?? "",
      attachments: [],
    };
  } catch (error) {
    logger.error("Error parsing email", error instanceof Error ? error.message : error);
    return null;
  }
}
